### PR TITLE
Adjust CL miner work batch size properly

### PR DESCRIPTION
**Pull request by LefterisJP**
We are now propery adjusting the batch size of the OpenCL miner
depending on the execution time of the last search.

### CODE REVIEW

File: libethash-cl/ethash_cl_miner.cpp [hunk 1]
```diff
@@ -495,19 +494,23 @@ void ethash_cl_miner::search(uint8_t const* header, uint64_t target, search_hook
 				pending.pop();
 			}
 
-/*			chrono::high_resolution_clock::duration d = chrono::high_resolution_clock::now() - t;
-			if (d > chrono::milliseconds(_msPerBatch * 10 / 9))
+			// adjust batch size depending on last search time
+			auto d = chrono::duration_cast<chrono::milliseconds>(chrono::high_resolution_clock::now() - t);
+			if (d != chrono::milliseconds(0)) // if duration is zero, we did not get in the actual search so adjust nothing
 			{
```
Comment (LefterisJP) [inline libethash-cl/ethash_cl_miner.cpp]: I know this goes against coding standards but as in many other places, this gives a clang warning for potential dangling If, so it asks you to explicitly add the braces.

**Comment by subtly** [comment]
Has this been tested on the GPUs we have in the office?

**Comment by subtly** [comment]
I'll followup my previous comment with -- I've tested this and it doesn't work as OpenCL is rejecting the altered work size (m_batchSize).

**Comment by LefterisJP** [comment]
Hm ... then this is new. I tested it in the GPU at the office and at the one I have home, and no rejection happened. We can limit the batch size by querying the limits of the batch size that OpenCL has (CL_DEVICE_MAX_WORK_GROUP_SIZE) and make sure any adjustments happens within those limits.

The problem with this is that I have learned not to trust OpenCL values as returned by `clGetDeviceInfo()`. But if as you say it can happen to a GPU that the batch size is rejected then this has to happen.

**Comment by LefterisJP** [comment]
rebased on develop

File: libethash-cl/ethash_cl_miner.cpp [hunk 2]
```diff
@@ -455,8 +455,7 @@ void ethash_cl_miner::search(uint8_t const* header, uint64_t target, search_hook
 		uint64_t start_nonce = uniform_int_distribution<uint64_t>()(engine);
 		for (;; start_nonce += m_batchSize)
 		{
-//			chrono::high_resolution_clock::time_point t = chrono::high_resolution_clock::now();
-
+			chrono::high_resolution_clock::time_point t = chrono::high_resolution_clock::now();
```
Comment (chfast) [inline libethash-cl/ethash_cl_miner.cpp]: use auto here too?

Comment (LefterisJP) [inline libethash-cl/ethash_cl_miner.cpp]: yeah why not, let me piggyback it in the next commit

**Comment by LefterisJP** [comment]
This Pull Request seems to be dealing with the batchSize, better known in OpenCL as the global work size. So let's make this be "The Fix".

I noticed that in [this](https://github.com/LefterisJP/cpp-ethereum/commit/a8eb96755cdffdc7e2c6f59003bc8cb9cc971166) commit @gavofyork changed it from `1024*256` to `1024*16` along with making a first attempt to introduce dynamic global work size adjustment. 

Miners in the mining channel have reported quite a dropped hashrate due to that specific change. I suggest 2 things:

1. Revert back to the original global work size
2. Make dynamic work size adjustment optional, off by default, and activateable by an option.

**Comment by Genoil** [comment]
Why try to adjust batch size on each search call? And why try to make it 100ms long? On most setups there's only one ideal batch size, that what gives the best hash rate. So instead of this you want some kind of auto tune that's present in a lot of crypto miners, that runs on first run by default and stores results in some settings file.

**Comment by gavofyork** [comment]
so we cannot abort during the course of a workload. this is what causes the large "pause" times (actually just the time spent waiting for the workload to finish and the abortion condition to be caught.

we should probably allow both workload specifying and a target max abort time (100ms in this patch) until we can instrument the best approach.

**Comment by Genoil** [comment]
Thanks for clarifying. On my GPU, 100ms is a ~2M workload that actually gives me the highest hashrate. A 1M workload (~50ms) gives approx. the same result, so there's quite some freedom. On Windows, there actually is a 2000ms abort time after which the driver resets itself. But I need to give the miner a 2 ^ 25 workload to crash it.

**Comment by LefterisJP** [comment]
rebased

**Comment by gavofyork** [comment]
fails on windows?

**Comment by LefterisJP** [comment]
It's all unrelated errors due to some develop failures yesterday. Will rebase again and see.

**Comment by LefterisJP** [comment]
Yet another rebase!

**Comment by LefterisJP** [comment]
This should resolve #2306 by providing better defaults and also allowing for customization of both global and local size.

**Comment by LefterisJP** [comment]
This should also resolve #2328 since we allow customization of the local work size (workgroup size)

**Comment by gavofyork** [comment]
fails to build on windows. header issue?

**Comment by LefterisJP** [comment]
Hmm I don't have a Windows build handy. @subtly do I recall correctly? This builds for you in Windows, right?

**Comment by LefterisJP** [comment]
I think I may know why the windows build failed. Let it run the tests again.

**Comment by LefterisJP** [comment]
Well at least now the windows build seems to pass. The OSX failures don't seem to have anything to do with me.

**Comment by gavofyork** [comment]
I'd prefer not to have horrible C constants in my nice C++ file.

**Comment by LefterisJP** [comment]
It's a bit more work to add constants as part of the class, and especially since they are exposed and used to also present the default values in the `--help` docstring. C macros are not so horrible. You also use them quite a bit.

But yes you are indeed correct that this should have been a c++ constant. Let me do that.

**Comment by conradjohnson** [comment]
I have 2 machines, both Ubuntu 14.04, one with AMD R9 290s and one with AMD 7990s.  I can test anything y'all want to throw at them on there if you want.  Been mining on testnet with ethminer for a while now.

**Comment by gavofyork** [comment]
seems the build fail is unrelated.

**Comment by conradjohnson** [comment]
Seems using multiple instances for multiple GPUs, it will drop the hash on the previous process.

**Comment by subtly** [comment]
This PR requires ethash-cl to link ethcore. Is that intentional?
This doesn't compile on macos and probably won't compile on windows.

File: libethcore/Ethash.cpp [hunk 3]
```diff
@@ -54,6 +54,9 @@ namespace dev
 namespace eth
 {
 
+const unsigned Ethash::defaultLocalWorkSize = 64;
```
Comment (subtly) [inline libethcore/Ethash.cpp]: How will ethash-cl resolve this value?

**Comment by LefterisJP** [comment]
@conradjohnson Can you elaborate? Are you by any chance in the Skype mining channel? We can chat about the issue there.

**Comment by conradjohnson** [comment]
Can you add me to skype mining channel?

**Pull request merged**